syntax = "proto3";

import "distribution.proto";
import "matrix.proto";

package bayesmix;

message NNIGState {
    double mean = 1;
    double var_scaling = 2;
    double shape = 3;
    double scale = 4;
}

message NNIGPrior {

  message NormalMeanPrior {
    UniNormalDistribution mean_prior = 1;
    double var_scaling = 2;
    double shape = 3;
    double scale = 4;
  }

  message NGGPrior {
    UniNormalDistribution mean_prior = 1;
    GammaDistribution var_scaling_prior = 2;
    double shape = 3;
    GammaDistribution scale_prior = 4;
  }

  oneof prior {
    NNIGState fixed_values = 1;
    NormalMeanPrior normal_mean_prior = 2;
    NGGPrior ngg_prior = 3;
  }
}


message LapNIGState {
    double mean = 1;
    double var = 2;
    double shape = 3;
    double scale = 4;
}

message LapNIGPrior {
  oneof prior {
    LapNIGState fixed_values = 1;
  }
}

message NNWState {
  Vector mean = 1;
  double var_scaling = 2;
  double deg_free = 3;
  Matrix scale = 4;
}

message NNWPrior {

  message NormalMeanPrior {
    MultiNormalDistribution mean_prior = 1;
    double var_scaling = 2;
    double deg_free = 3;
    Matrix scale = 4;
  }

  message NGIWPrior {
    MultiNormalDistribution mean_prior = 1;
    GammaDistribution var_scaling_prior = 2;
    double deg_free = 3;
    InvWishartDistribution scale_prior = 4;
  }

  oneof prior {
    NNWState fixed_values = 1;
    NormalMeanPrior normal_mean_prior = 2;
    NGIWPrior ngiw_prior = 3;
  }
}


message LinRegUniState {
  Vector mean = 1;
  Matrix var_scaling = 2;
  double shape = 3;
  double scale = 4;
}


message LinRegUniPrior {
  oneof prior {
    LinRegUniState fixed_values = 1;
  }
}

message EmptyPrior {
  double fake_field = 1;
}
